cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519ML.hs

```diff
@@ -0,0 +1,278 @@
+-- import qualified Data.ByteString.Unsafe as BS
+import Data.Proxy
+import Control.Monad.Class.MonadThrow (MonadThrow (..), throwIO)
+import Control.Monad.Class.MonadST (MonadST (..))
+import Control.Monad.ST (ST)
+import Control.Monad.ST.Unsafe (unsafeIOToST)
+
+import Cardano.Binary (FromCBOR (..), ToCBOR (..))
+
+import Cardano.Foreign
+import Cardano.Crypto.PinnedSizedBytes
+import Cardano.Crypto.Libsodium.C
+import Cardano.Crypto.Libsodium (MLockedSizedBytes)
+import Cardano.Crypto.MonadSodium
+  ( MonadSodium (..)
+  , mlsbToByteString
+  , mlsbFromByteStringCheck
+  , mlsbUseAsSizedPtr
+  , mlsbNew
+  , mlsbFinalize
+  )
+
+import Cardano.Crypto.DSIGNM.Class
+import Cardano.Crypto.MLockedSeed
+import Cardano.Crypto.Util (SignableRepresentation(..))
+
+data Ed25519DSIGNM
+
+instance NoThunks (VerKeyDSIGNM Ed25519DSIGNM)
+instance NoThunks (SignKeyDSIGNM Ed25519DSIGNM)
+instance NoThunks (SigDSIGNM Ed25519DSIGNM)
+
+-- | Convert C-style return code / errno error reporting into Haskell
+-- exceptions.
+--
+-- Runs an IO action (which should be some FFI call into C) that returns a
+-- result code; if the result code returned is nonzero, fetch the errno, and
+-- return it.
+cOrError :: MonadST m => (forall s. ST s Int) -> m (Maybe Errno)
+cOrError action = do
+  withLiftST $ \fromST -> fromST $ do
+    res <- action
+    if (res == 0) then
+      return Nothing
+    else
+      Just <$> unsafeIOToST getErrno
+
+-- | Throws an appropriate 'IOException' when 'Just' an 'Errno' is given.
+throwOnErrno :: (MonadThrow m) => String -> String -> Maybe Errno -> m ()
+throwOnErrno contextDesc cFunName maybeErrno = do
+  case maybeErrno of
+    Just errno -> throwIO $ errnoToIOError (contextDesc ++ ": " ++ cFunName) errno Nothing Nothing
+    Nothing -> return ()
+
+instance DSIGNMAlgorithmBase Ed25519DSIGNM where
+    type SeedSizeDSIGNM Ed25519DSIGNM = CRYPTO_SIGN_ED25519_SEEDBYTES
+    -- | Ed25519 key size is 32 octets
+    -- (per <https://tools.ietf.org/html/rfc8032#section-5.1.6>)
+    type SizeVerKeyDSIGNM  Ed25519DSIGNM = CRYPTO_SIGN_ED25519_PUBLICKEYBYTES
+    -- | Ed25519 secret key size is 32 octets; however, libsodium packs both
+    -- the secret key and the public key into a 64-octet compound and exposes
+    -- that as the secret key; the actual 32-octet secret key is called
+    -- \"seed\" in libsodium. For backwards compatibility reasons and
+    -- efficiency, we use the 64-octet compounds internally (this is what
+    -- libsodium expects), but we only serialize the 32-octet secret key part
+    -- (the libsodium \"seed\").
+    type SizeSignKeyDSIGNM Ed25519DSIGNM = CRYPTO_SIGN_ED25519_SEEDBYTES
+    -- | Ed25519 signature size is 64 octets
+    type SizeSigDSIGNM     Ed25519DSIGNM = CRYPTO_SIGN_ED25519_BYTES
+
+    --
+    -- Key and signature types
+    --
+
+    newtype VerKeyDSIGNM Ed25519DSIGNM = VerKeyEd25519DSIGNM (PinnedSizedBytes (SizeVerKeyDSIGNM Ed25519DSIGNM))
+        deriving (Show, Eq, Generic)
```

Comment:
Good point. However, this only holds for the sign keys; ver keys and signatures are not stored in mlocked memory, because they are public information anyway. So I'm keeping the generic instances for those, and just provide custom instances for sign keys and seeds (where applicable).

Reply:
Oh yeah, sorry. :+1:  Didn't pay enough attention and mixed up VerKey and SignKey